### PR TITLE
Leaves_Ga-Young

### PR DESCRIPTION
**Pull request by gyjin**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | I created a "self.top_1" method in the work model to do the logic of finding the work that had the most votes. 
Describe how you approached testing that model method. What edge cases did you come up with? | Because I know which work has the most votes from the fixtures, I can test that the return of calling that method will be the work that I expect. Edge cases would be if there are not works or if there are more than 1 highest voted work. 
What are `session` and `flash`? What is the difference between them? | Flash is used to send one-time messages from controllers to views. They are usually used to confirm to the human user that an action was successful or failed. Session is an object that holds data throughout a "session" on the browser, which usually ends when the browser is closed. This can be used to keep track of a user logged-in into a site. 
What was one thing that you gained more clarity on through this assignment? | I learned that different models can be connected by nested routes. I also learned more about whether a certain method belongs in the model or controller. 
What is the Heroku URL of your deployed application? | https://gyjin-media-ranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | I thought this project went overall pretty smoothly. I don't have any pressing issues to bring up.

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+Rails.application.routes.draw do
+  root 'homepages#index'
+  resources :homepages
+  resources :works
+  resources :votes
+  resources :users
```
Comment (dHelmgren) [inline config/routes.rb:6]: You aren't using all of these.

File: app/views/works/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<h2><strong>List of Works</strong></h2>
+
+<%= render partial: "print_all", locals: {category_name: "Albums", method_name: @all_albums} %>
+
+<%= render partial: "print_all", locals: {category_name: "Books", method_name: @all_books} %>
+
+<%= render partial: "print_all", locals: {category_name: "Movies", method_name: @all_movies} %>
```
Comment (dHelmgren) [inline app/views/works/index.html.erb:7]: YEAH!

File: app/models/work.rb [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+      end
+    end
+    sorted_books = books.sort {|a,b| b.votes.count <=> a.votes.count}
+    return sorted_books
+  end
+  
+  def self.all_movies
+    movies = []
+    Work.all.each do |work|
+      if work.category == "movie"
```
Comment (dHelmgren) [inline app/models/work.rb:35]: These methods share a lot of code. Is there a way you could dry this up?

File: app/controllers/votes_controller.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  end
+  
+  def edit
+  end
+  
+  def update
+  end
+  
+  def destroy
+  end
```
Comment (dHelmgren) [inline app/controllers/votes_controller.rb:15]: This should be a hint that this code isn't doing anything for you.

File: app/controllers/works_controller.rb [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+      flash[:warning] = "A problem occured: You must log in to do that"
+      return
+    elsif @user
+      @user.votes.each do |vote|
+        if vote.work_id == @work.id
+          redirect_to work_path(@work.id)
+          flash[:warning] = "A problem occured: Could not upvote. User has already voted for this media."
+          return
+        end
+      end
```
Comment (dHelmgren) [inline app/controllers/works_controller.rb:72]: This might benefit from a controller action. You can apply that filter from Application controller

**Comment by dHelmgren** [comment]
# Media Ranker
## What We're Looking For



### Manual Testing

Workflow | yes / no
--- | ---
Deployed to Heroku | yes
Optional: Look and feel is similar to the original (consider styling and layout) | yes
**Before logging in** |
On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | yes
Can go into a work's show page | yes
Verify unable to vote on a work, and get a flash message | yes
Can edit this work successfully, and get a flash message | yes
Can go to "View all media" page and see three lists of works, sorted by vote | Yes, but can't tell yet
Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | yes
Can create a new work successfully. Note the URL for this work's show page | yes
Can delete this work successfully | yes
Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | Head Only
Verify that the "View all users" page lists no users (except what is produced by db seeding) | yes
**Log in** |
Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | yes
Your username is listed in "View all users" page | yes
Verify that number of votes determines the Media Spotlight | yes
Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | yes
Voting on the same work twice produces an error/flash message, and there is no extra vote | yes
**Log out** |
Logging out showed a flash message and changed the UI | yes
Logging in as a new user creates a new user | yes
Logging in as an already existing user has a specific flash message | yes

### Targeted Code Review

Area | yes / no
--- | ---
git commits were small and atomic, with useful messages | yes
Routes file uses `resources` for works | yes
Routes file shows intention in limiting routes for voting, log-in functionality, and users | no, see comment
The homepage view, all media view, and new works view use semantic HTML | yes
The homepage view, all media view, and new works view use partials when appropriate | yes!
The model for media (likely named `work.rb`) `has_many` votes | yes
The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | yes, though see comments.
Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | yes
Some controller, likely the `WorksController`, has a controller filter for finding a work | yes
The `WorksController` uses strong params | yes
The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | yes

#### Targed Test Review

Area | yes / no
--- | ---
There are valid fixtures files used for users, votes, and works | yes
User model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | yes
Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Work model has tests with a section on all business logic methods in the model, including their edge cases | yes

## Overall Feedback
Good job overall! I have a couple of cheers and jeers for different things in this project, but it's clear the learning goals were met.
